clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -679,10 +679,11 @@ private long calcCompletionTimesAndReturnSessionLifetimeMs() {
                 else if (connectionsMaxReauthMs == null)
                     retvalSessionLifetimeMs = zeroIfNegative(credentialExpirationMs - authenticationEndMs);
                 else
-                    retvalSessionLifetimeMs = zeroIfNegative(
-                            Math.min(credentialExpirationMs - authenticationEndMs, connectionsMaxReauthMs));
+                    retvalSessionLifetimeMs = zeroIfNegative(Math.min(credentialExpirationMs - authenticationEndMs, connectionsMaxReauthMs));
 
-                sessionExpirationTimeNanos = authenticationEndNanos + 1000 * 1000 * retvalSessionLifetimeMs;
+                if (connectionsMaxReauthMs != null) {
```

Comment:
>when reauth is disabled (when max reauth ms is NOT set), leave ReauthInfo#sessionExpirationTimeNanos as null but return millis until the token expires in SaslAuthenticateResponse's sessionLifetimeMs

I think I have a different understanding. My mental model was that if either of `credentialExpirationMs != null || connectionsMaxReauthMs != null` was `true` we have re-auth enabled. The re-auth disabled case I was thinking of was when both were `false`. Now that I think about that I'm not sure that is a valid scenario.

I think your original change to make setting `sessionExpirationTimeNanos` un-conditional was the right one, my `>=0` test is effectively the same as making it un-conditional. Therefore my suggestion of delayed initialisation becomes moot as we don't need to worry about excluding the default value from the test.
I was wondering if the property change was what was actually causing the test failures that you found rather than making it un-conditional.

Reply:
@showuon what is your thinking on this?

Reply:
> I was wondering if the property change was what was actually causing the test failures that you found rather than making it un-conditional.

@SamBarker ,I don't understand this. Which property are you referring to?

Reply:
OK, for @SamBarker , I think we can discuss the property change in a separate thread since it's related to tests.

For this change:
```
if (connectionsMaxReauthMs != null) {
       sessionExpirationTimeNanos = authenticationEndNanos + 1000 * 1000 * retvalSessionLifetimeMs;
}
```
I understand why you did this @acsaki . It's because you think:

> when reauth is disabled (when max reauth ms is NOT set), leave ReauthInfo#sessionExpirationTimeNanos as null but return millis until the token expires in SaslAuthenticateResponse's sessionLifetimeMs

I think it's correct, **IF** the sasl client did close the connection after the `sessionLifetimeMs`. But I don't think we should have this optimistic assumption for this "potential" security issue. I agree with @SamBarker about your original version of "removing the if condition" is a good fix.:
```
...
else
      retvalSessionLifetimeMs = zeroIfNegative(Math.min(credentialExpirationMs - authenticationEndMs, 

sessionExpirationTimeNanos = authenticationEndNanos + 1000 * 1000 * retvalSessionLifetimeMs;
```

That is, no matter the reauth is enabled or not, we set the `sessionExpirationTimeNanos`, to inform the server, too. So that we can make sure when the session expired, either server or client will kill this connection. WDYT?

Reply:
Also, the debug line in L698 confused me a little. When `credentialExpirationMs != null && sessionExpirationTimeNanos == null`, we'll log:
```
"Authentication complete; session max lifetime from broker config={} ms, credential expiration={} ({} ms); no session expiration, sending 0 ms to client"
```
The point I'm confused here is `no session expiration`. Why no session expiration? Since `credentialExpirationMs != null`, there must be some credential expiration time, so in this case, sending 0 to client doesn't make sense to me, right? I think the log should also be updated. WDYT?

Reply:
@rondagostino @rajinisivaram , do you have any comments? Thanks.

Reply:
@showuon this is what I've also found rather confusing. I agree that after the token expires the connection should be closed sooner or later which isn't going to happen when sessionExpirationTimeNanos is not set. While there is the Authenticator interface where comments suggest that  #serverSessionExpirationTimeNanos should be left as null when re-authentication is "disabled". Does it make sense for reauth to be disabled? Or rather there are some clients or SASL mechanisms where we don't expect reauthentication to ever happen?

Reply:
> While there is the Authenticator interface where comments suggest that #serverSessionExpirationTimeNanos should be left as null when re-authentication is "disabled"

Is [this](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/network/Authenticator.java#L90-L102) javadoc what you mean? From the javadoc, I don't see it said we should return null for re-auth disabled. I think it's OK to return the value when re-auth disabled.

> here are some clients or SASL mechanisms where we don't expect reauthentication to ever happen?

Yes, it is. You can check [this](https://github.com/edenhill/librdkafka/issues/3304) for reference.

Thanks.